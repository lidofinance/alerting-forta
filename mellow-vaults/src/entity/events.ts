import { FindingSeverity, FindingType, Log } from 'forta-agent'

export type EventOfNotice = {
  name: string
  address: string
  event: string
  alertId: string
  description: CallableFunction
  severity: FindingSeverity
  type?: FindingType
  uniqueKey?: string
}

export type TransactionDtoTx = {
  data: string
  hash: string
}

export type TransactionDto = {
  logs: Log[]
  to: string | null
  block: {
    timestamp: number
    number: number
  }
  transaction: TransactionDtoTx
}

export type BlockDto = {
  number: number
  timestamp: number
  parentHash: string
}
