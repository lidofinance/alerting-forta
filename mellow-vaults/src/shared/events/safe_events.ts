import { FindingSeverity } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { SafeTX } from 'constants/common'
import { getSafeLink, getSafeTxLink, getTxLink } from '../string'

export const GNOSIS_SAFE_EVENTS_OF_NOTICE = [
  {
    event: 'event AddedOwner(address owner)',
    alertId: 'SAFE-OWNER-ADDED',
    name: '🚨 Gnosis Safe: Owner added',
    description: (safeTx: SafeTX, args: Result) => `New owner ${args.owner} was added to ${getSafeLink(safeTx)}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event RemovedOwner(address owner)',
    alertId: 'SAFE-OWNER-REMOVED',
    name: '🚨 Gnosis Safe: Owner removed',
    description: (safeTx: SafeTX, args: Result) => `Owner ${args.owner} was removed from ${getSafeLink(safeTx)}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event ChangedFallbackHandler(address handler)',
    alertId: 'SAFE-HANDLER-CHANGED',
    name: '🚨 Gnosis Safe: Fallback handler changed',
    description: (safeTx: SafeTX, args: Result) =>
      `Fallback handler for ${getSafeLink(safeTx)} ` + `was changed to ${args.handler}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event ChangedGuard(address guard)',
    alertId: 'SAFE-GUARD-CHANGED',
    name: '🚨 Gnosis Safe: Guard changed',
    description: (safeTx: SafeTX, args: Result) => `Guard for ${getSafeLink(safeTx)} was changed to ${args.guard}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event ChangedThreshold(uint256 threshold)',
    alertId: 'SAFE-THRESHOLD-CHANGED',
    name: '🚨 Gnosis Safe: Threshold changed',
    description: (safeTx: SafeTX, args: Result) =>
      `Threshold for ${getSafeLink(safeTx)} was changed to ${args.threshold}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event DisabledModule(address module)',
    alertId: 'SAFE-MODULE-DISABLED',
    name: '🚨 Gnosis Safe: Module disabled',
    description: (safeTx: SafeTX, args: Result) => `Module ${args.module} was disabled for ${getSafeLink(safeTx)}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event EnabledModule(address module)',
    alertId: 'SAFE-MODULE-ENABLED',
    name: '🚨 Gnosis Safe: Module enabled',
    description: (safeTx: SafeTX, args: Result) => `Module ${args.module} was enabled for ${getSafeLink(safeTx)}`,
    severity: FindingSeverity.Medium,
  },
  {
    event: 'event ExecutionFailure(bytes32 txHash, uint256 payment)',
    alertId: 'SAFE-EXECUTION-FAILURE',
    name: '❌ Gnosis Safe: TX Execution failed',
    description: (safeTx: SafeTX, args: Result) =>
      `[TX](${getSafeTxLink(safeTx)}) execution failed for ` +
      `${getSafeLink(safeTx)}\n` +
      `[blockchain explorer](${getTxLink(safeTx)})`,
    severity: FindingSeverity.Info,
  },
  {
    event: 'event ExecutionSuccess(bytes32 txHash, uint256 payment)',
    alertId: 'SAFE-EXECUTION-SUCCESS',
    name: '✅ Gnosis Safe: TX Executed',
    description: (safeTx: SafeTX, args: Result) =>
      `[TX](${getSafeTxLink(safeTx)}) executed by ${getSafeLink(safeTx)}\n` +
      `[blockchain explorer](${getTxLink(safeTx)})`,
    severity: FindingSeverity.Info,
  },
  {
    event: 'event ExecutionFromModuleFailure(address module)',
    alertId: 'SAFE-EXECUTION-FAILURE-FROM-MODULE',
    name: '❌ Gnosis Safe: Execution failed from module',
    description: (safeTx: SafeTX, args: Result) =>
      `TX execution failed for ${getSafeLink(safeTx)} ` + `from module ${args.module}`,
    severity: FindingSeverity.Info,
  },
  {
    event: 'event ExecutionFromModuleSuccess(address module)',
    alertId: 'SAFE-EXECUTION-SUCCESS-FROM-MODULE',
    name: '✅ Gnosis Safe: Execution success from module',
    description: (safeTx: SafeTX, args: Result) =>
      `Execution success for ${getSafeLink(safeTx)} from module ${args.module}`,
    severity: FindingSeverity.Info,
  },
]
