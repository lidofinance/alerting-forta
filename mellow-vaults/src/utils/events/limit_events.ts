import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'

export const MELLOW_VAULT_INCREASE_LIMIT_EVENT = 'event IncreaseLimit(uint256 amount)'

export const vaultLimitNotice = {
  IncreaseLimit: {
    event: MELLOW_VAULT_INCREASE_LIMIT_EVENT,
    alertId: 'MELLOW-SYMBIOTIC-INCREASE-LIMIT',
    name: '⚠️ Vault: Symbiotic limit increased',
    description: (args: Result) => `Symbiotic limit increased to ${args.amount}`,
    severity: FindingSeverity.Medium,
    type: FindingType.Info,
    uniqueKey: '27c9ac49-4c5f-4d3e-9a4a-0d61e21965b2 ',
  },
}
