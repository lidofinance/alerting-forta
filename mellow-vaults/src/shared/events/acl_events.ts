import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { ACL_ROLES } from 'constants/common'

export function getACLEvents(DEFAULT_BOND_STRATEGY_ADDRESS: string): EventOfNotice[] {
  const uniqueKeys = [
    '12c18264-900b-48a9-bdee-b193b61a36c6',
    'd45a13a5-940c-439f-96e2-2cec78f9de24',
    '19ba7243-ff94-45af-aa3d-6b12a7f0b83d',
  ]

  return [
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event:
        'event RoleAdminChanged(bytes32 indexed role, bytes32 indexed previousAdminRole, bytes32 indexed newAdminRole)',
      alertId: 'MELLOW-VAULT-ROLE-ADMIN-CHANGED',
      name: 'ðŸš¨ Vault: Role Admin changed',
      description: (args: Result) =>
        `Role Admin for role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
        `was changed from ${args.previousAdminRole} to ${args.newAdminRole} ` +
        `bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[0],
    },
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event RoleGranted(bytes32 indexed role, address indexed account, address indexed sender)',
      alertId: 'MELLOW-VAULT-ROLE-GRANTED',
      name: 'ðŸš¨ Vault: Role granted',
      description: (args: Result) =>
        `Role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
        `was granted to ${args.account} by ${args.sender} ` +
        `bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[1],
    },
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event RoleRevoked(bytes32 indexed role, address indexed account, address indexed sender)',
      alertId: 'MELLOW-VAULT-ROLE-REVOKED',
      name: 'ðŸš¨  Vault: Role revoked',
      description: (args: Result) =>
        `Role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
        `was revoked to ${args.account} by ${args.sender} ` +
        `bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[2],
    },
  ]
}
