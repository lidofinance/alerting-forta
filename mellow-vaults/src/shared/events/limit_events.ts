import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { MELLOW_SYMBIOTIC_ADDRESS } from 'constants/common'

export function getLimitEvents(): EventOfNotice[] {
  const uniqueKeys = ['27c9ac49-4c5f-4d3e-9a4a-0d61e21965b2 ']

  return [
    {
      address: MELLOW_SYMBIOTIC_ADDRESS,
      event: 'event IncreaseLimit(uint256 amount)',
      alertId: 'MELLOW-SYMBIOTIC-INCREASE-LIMIT',
      name: '⚠️ Vault: Symbiotic limit increased',
      description: (args: Result) => `Symbiotic limit increased to ${args.amount}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[0],
    },
  ]
}
