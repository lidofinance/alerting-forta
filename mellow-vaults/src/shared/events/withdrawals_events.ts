import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { VAULT_LIST } from 'constants/common'

export function getProcessAllEvents(DEFAULT_BOND_STRATEGY_ADDRESS: string): EventOfNotice[] {
  const uniqueKeys = ['9d846c64-6858-47c4-afd2-3f3d31da079b']

  const vault = VAULT_LIST.find((vault) => vault.defaultBondStrategy === DEFAULT_BOND_STRATEGY_ADDRESS)
  return [
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event DefaultBondStrategyProcessWithdrawals (address[] users, uint256 timestamp)',
      alertId: 'MELLOW-VAULT-WITHDRAWAL-ALL',
      name: 'ℹ️ Vault: Withdrawal all',
      description: () => {
        return `Mellow Vault [${vault?.name}] (${vault?.vault}) - withdrawal, bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`
      },
      severity: FindingSeverity.Info,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[0],
    },
  ]
}

export function getProcessWithdrawalsEvents(DEFAULT_BOND_STRATEGY_ADDRESS: string): EventOfNotice[] {
  const uniqueKeys = ['e27d8c35-bce0-4901-a3d6-ca232cf68ede']

  const vault = VAULT_LIST.find((vault) => vault.defaultBondStrategy === DEFAULT_BOND_STRATEGY_ADDRESS)
  return [
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event DefaultBondStrategyProcessWithdrawals (address[] users, uint256 timestamp)',
      alertId: 'MELLOW-VAULT-PARTIAL-WITHDRAWAL',
      name: '⚠️ Vault: Withdrawal partial',
      description: () => {
        return `Mellow Vault [${vault?.name}] (${vault?.vault}) - bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`
      },
      severity: FindingSeverity.Medium,
      type: FindingType.Suspicious,
      uniqueKey: uniqueKeys[0],
    },
  ]
}
