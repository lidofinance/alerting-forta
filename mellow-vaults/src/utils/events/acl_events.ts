import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { ACL_ROLES, VAULT_LIST } from 'constants/common'

const MELLOW_VAULT_STRATEGY_ROLE_ADMIN_CHANGED_EVENT =
  'event RoleAdminChanged(bytes32 indexed role, bytes32 indexed previousAdminRole, bytes32 indexed newAdminRole)'

const MELLOW_VAULT_STRATEGY_ROLE_GRANTED_EVENT =
  'event RoleGranted(bytes32 indexed role, address indexed account, address indexed sender)'

const MELLOW_VAULT_STRATEGY_ROLE_REVOKED_EVENT =
  'event RoleRevoked(bytes32 indexed role, address indexed account, address indexed sender)'

export const MELLOW_VAULT_STRATEGY_ACL_EVENTS = [
  MELLOW_VAULT_STRATEGY_ROLE_ADMIN_CHANGED_EVENT,
  MELLOW_VAULT_STRATEGY_ROLE_GRANTED_EVENT,
  MELLOW_VAULT_STRATEGY_ROLE_REVOKED_EVENT,
]

const getBondStrategyNameByAddress = (address: string) =>
  VAULT_LIST.find((vault) => vault.defaultBondStrategy === address)

export const aclNotices: Record<string, EventOfNotice> = {
  RoleAdminChanged: {
    event: MELLOW_VAULT_STRATEGY_ROLE_ADMIN_CHANGED_EVENT,
    alertId: 'MELLOW-VAULT-ROLE-ADMIN-CHANGED',
    name: 'ðŸš¨ Vault: Role Admin changed',
    description: (args: Result, address: string) =>
      `Mellow Vault [${getBondStrategyNameByAddress(address)?.name}] bond strategy ` +
      `Role Admin for role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
      `was changed from ${args.previousAdminRole} to ${args.newAdminRole} ` +
      `bond strategy address - ${address}`,
    severity: FindingSeverity.High,
    type: FindingType.Info,
    uniqueKey: '12c18264-900b-48a9-bdee-b193b61a36c6',
  },
  RoleGranted: {
    event: MELLOW_VAULT_STRATEGY_ROLE_GRANTED_EVENT,
    alertId: 'MELLOW-VAULT-ROLE-GRANTED',
    name: 'ðŸš¨ Vault: Role granted',
    description: (args: Result, address: string) =>
      `Mellow Vault [${getBondStrategyNameByAddress(address)?.name}] bond strategy ` +
      `Role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
      `was granted to ${args.account} by ${args.sender} ` +
      `bond strategy address - ${address}`,
    severity: FindingSeverity.High,
    type: FindingType.Info,
    uniqueKey: 'd45a13a5-940c-439f-96e2-2cec78f9de24',
  },
  RoleRevoked: {
    event: MELLOW_VAULT_STRATEGY_ROLE_REVOKED_EVENT,
    alertId: 'MELLOW-VAULT-ROLE-REVOKED',
    name: 'ðŸš¨  Vault: Role revoked',
    description: (args: Result, address: string) =>
      `Mellow Vault [${getBondStrategyNameByAddress(address)?.name}] bond strategy ` +
      `Role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
      `was revoked to ${args.account} by ${args.sender} ` +
      `bond strategy address - ${address}`,
    severity: FindingSeverity.High,
    type: FindingType.Info,
    uniqueKey: '19ba7243-ff94-45af-aa3d-6b12a7f0b83d',
  },
}
