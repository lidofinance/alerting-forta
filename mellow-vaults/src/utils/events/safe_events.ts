import { EventOfNotice, TransactionDtoTx } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { getSafeLink, getSafeTxLink, getTxLink } from '../../shared/string'
import { MELLOW_VAULT_ADMIN_MULTISIGS } from 'constants/common'

const MSIG_ADDED_OWNER_EVENT = 'event AddedOwner(address owner)'
const MSIG_REMOVED_OWNER_EVENT = 'event RemovedOwner(address owner)'
const MSIG_CHANGED_FALLBACK_HANDLER_EVENT = 'event ChangedFallbackHandler(address handler)'
const MSIG_CHANGED_GUARD_EVENT = 'event ChangedGuard(address guard)'
const MSIG_CHANGED_THRESHOLD_EVENT = 'event ChangedThreshold(uint256 threshold)'
const MSIG_DISABLED_MODULE_EVENT = 'event DisabledModule(address module)'
const MSIG_ENABLED_MODULE_EVENT = 'event EnabledModule(address module)'
const MSIG_EXECUTION_FAILURE_EVENT = 'event ExecutionFailure(bytes32 txHash, uint256 payment)'
const MSIG_EXECUTION_SUCCESS_EVENT = 'event ExecutionSuccess(bytes32 txHash, uint256 payment)'
const MSIG_EXECUTION_FROM_MODULE_FAILURE_EVENT = 'event ExecutionFromModuleFailure(address module)'
const MSIG_EXECUTION_FROM_MODULE_SUCCESS_EVENT = 'event ExecutionFromModuleSuccess(address module)'

export const MSIG_EVENTS = [
  MSIG_ADDED_OWNER_EVENT,
  MSIG_REMOVED_OWNER_EVENT,
  MSIG_CHANGED_FALLBACK_HANDLER_EVENT,
  MSIG_CHANGED_GUARD_EVENT,
  MSIG_CHANGED_THRESHOLD_EVENT,
  MSIG_DISABLED_MODULE_EVENT,
  MSIG_ENABLED_MODULE_EVENT,
  MSIG_EXECUTION_FAILURE_EVENT,
  MSIG_EXECUTION_SUCCESS_EVENT,
  MSIG_EXECUTION_FROM_MODULE_FAILURE_EVENT,
  MSIG_EXECUTION_FROM_MODULE_SUCCESS_EVENT,
]

const safes = MELLOW_VAULT_ADMIN_MULTISIGS

const safeTx = (args: Result, address: string, tx: TransactionDtoTx) => {
  const [safeAddress, safeName] = safes.find((item) => item[0] === address) ?? []
  return {
    tx: tx.hash,
    safeAddress,
    safeName,
    safeTx: args.txHash || '',
  }
}

export const safeNotices: Record<string, EventOfNotice> = {
  AddedOwner: {
    address: '',
    event: MSIG_ADDED_OWNER_EVENT,
    alertId: 'SAFE-OWNER-ADDED',
    name: '🚨 Gnosis Safe: Owner added',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `New owner ${args.owner} was added to ${getSafeLink(safeTx(args, address, tx))}`,
    severity: FindingSeverity.Medium,
    type: FindingType.Info,
    uniqueKey: '14f4c38e-7e1e-49df-af02-6c582e79158c',
  },
  RemovedOwner: {
    address: '',
    event: MSIG_REMOVED_OWNER_EVENT,
    alertId: 'SAFE-OWNER-REMOVED',
    name: '🚨 Gnosis Safe: Owner removed',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `Owner ${args.owner} was removed from ${getSafeLink(safeTx(args, address, tx))}`,
    severity: FindingSeverity.Medium,
    type: FindingType.Info,
    uniqueKey: '6a6b7566-b398-49f7-95d0-a8353fe53429',
  },
  ChangedFallbackHandler: {
    address: '',
    event: MSIG_CHANGED_FALLBACK_HANDLER_EVENT,
    alertId: 'SAFE-HANDLER-CHANGED',
    name: '🚨 Gnosis Safe: Fallback handler changed',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `Fallback handler for ${getSafeLink(safeTx(args, address, tx))} ` + `was changed to ${args.handler}`,
    severity: FindingSeverity.Medium,
    type: FindingType.Info,
    uniqueKey: '8c4fda92-35ea-4370-8bc8-9d9f952859f7',
  },
  ChangedGuard: {
    address: '',
    event: MSIG_CHANGED_GUARD_EVENT,
    alertId: 'SAFE-GUARD-CHANGED',
    name: '🚨 Gnosis Safe: Guard changed',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `Guard for ${getSafeLink(safeTx(args, address, tx))} was changed to ${args.guard}`,
    severity: FindingSeverity.Medium,
    type: FindingType.Info,
    uniqueKey: '1af528e5-d397-437c-ab19-0921c1443b3f',
  },
  ChangedThreshold: {
    address: '',
    event: MSIG_CHANGED_THRESHOLD_EVENT,
    alertId: 'SAFE-THRESHOLD-CHANGED',
    name: '🚨 Gnosis Safe: Threshold changed',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `Threshold for ${getSafeLink(safeTx(args, address, tx))} was changed to ${args.threshold}`,
    severity: FindingSeverity.Medium,
    type: FindingType.Info,
    uniqueKey: 'e70d0bb5-0033-4f34-a1cc-3b9f177f215f',
  },
  DisabledModule: {
    address: '',
    event: MSIG_DISABLED_MODULE_EVENT,
    alertId: 'SAFE-MODULE-DISABLED',
    name: '🚨 Gnosis Safe: Module disabled',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `Module ${args.module} was disabled for ${getSafeLink(safeTx(args, address, tx))}`,
    severity: FindingSeverity.Medium,
    type: FindingType.Info,
    uniqueKey: 'cbfb31ed-b0e1-4007-a2ae-940de626ff2e',
  },
  EnabledModule: {
    address: '',
    event: MSIG_ENABLED_MODULE_EVENT,
    alertId: 'SAFE-MODULE-ENABLED',
    name: '🚨 Gnosis Safe: Module enabled',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `Module ${args.module} was enabled for ${getSafeLink(safeTx(args, address, tx))}`,
    severity: FindingSeverity.Medium,
    type: FindingType.Info,
    uniqueKey: 'cc404aa8-a31e-4fce-9c59-8f0b2a5d41a5',
  },
  ExecutionFailure: {
    address: '',
    event: MSIG_EXECUTION_FAILURE_EVENT,
    alertId: 'SAFE-EXECUTION-FAILURE',
    name: '❌ Gnosis Safe: TX Execution failed',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `[TX](${getSafeTxLink(safeTx(args, address, tx))}) execution failed for ` +
      `${getSafeLink(safeTx(args, address, tx))}\n` +
      `[blockchain explorer](${getTxLink(safeTx(args, address, tx))})`,
    severity: FindingSeverity.Info,
    type: FindingType.Info,
    uniqueKey: '15587c71-32ce-44ea-8232-67676f3ba848',
  },
  ExecutionSuccess: {
    address: '',
    event: MSIG_EXECUTION_SUCCESS_EVENT,
    alertId: 'SAFE-EXECUTION-SUCCESS',
    name: '✅ Gnosis Safe: TX Executed',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `[TX](${getSafeTxLink(safeTx(args, address, tx))}) executed by ${getSafeLink(safeTx(args, address, tx))}\n` +
      `[blockchain explorer](${getTxLink(safeTx(args, address, tx))})`,
    severity: FindingSeverity.Info,
    type: FindingType.Info,
    uniqueKey: '8dd6c439-842b-4a2a-a7e8-b404ffbac5ce',
  },
  ExecutionFromModuleFailure: {
    address: '',
    event: MSIG_EXECUTION_FROM_MODULE_FAILURE_EVENT,
    alertId: 'SAFE-EXECUTION-FAILURE-FROM-MODULE',
    name: '❌ Gnosis Safe: Execution failed from module',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `TX execution failed for ${getSafeLink(safeTx(args, address, tx))} ` + `from module ${args.module}`,
    severity: FindingSeverity.Info,
    type: FindingType.Info,
    uniqueKey: 'aaf17ef8-ea92-47f6-8663-c11b0c61a1c6',
  },
  ExecutionFromModuleSuccess: {
    address: '',
    event: MSIG_EXECUTION_FROM_MODULE_SUCCESS_EVENT,
    alertId: 'SAFE-EXECUTION-SUCCESS-FROM-MODULE',
    name: '✅ Gnosis Safe: Execution success from module',
    description: (args: Result, address: string, tx: TransactionDtoTx) =>
      `Execution success for ${getSafeLink(safeTx(args, address, tx))} from module ${args.module}`,
    severity: FindingSeverity.Info,
    type: FindingType.Info,
    uniqueKey: '801cb979-5b3f-42b3-a471-4195112b2fd2',
  },
}
