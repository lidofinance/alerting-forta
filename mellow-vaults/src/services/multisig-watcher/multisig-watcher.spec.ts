import { TransactionDto } from '../../entity/events'
import * as Winston from 'winston'
import { ethers, Finding } from 'forta-agent'

import { MultisigWatcherSrv } from './MultisigWatcher.srv'

const TEST_TIMEOUT = 120_000 // ms

describe('MultisigWatcher srv functional tests', () => {
  const drpcURL = `https://eth.drpc.org`
  const mainnet = 1
  const ethProvider = new ethers.providers.JsonRpcProvider(drpcURL, mainnet)

  const logger: Winston.Logger = Winston.createLogger({
    format: Winston.format.simple(),
    transports: [new Winston.transports.Console()],
  })

  const service = new MultisigWatcherSrv(logger)

  let runTransaction: (txHash: string, initBlock?: number) => Promise<Finding[]>

  beforeAll(() => {
    jest.spyOn(console, 'log').mockImplementation(() => {})
    jest.spyOn(Date, 'now').mockImplementation(() => new Date('2024-01-01').getTime())

    runTransaction = async (txHash: string) => {
      const trx = await ethProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        transaction: {
          data: trx.data || '',
          hash: trx.hash || '',
        },
      }

      return service.handleTransaction(transactionDto)
    }
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  it(
    'should process tx with Added Owner and Execution Success',
    async () => {
      const findings = await runTransaction('0x4c3e507fd78daac24dd75e33e7c69381481db8bce62c795d2eec67126ad6d396')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
  it(
    'should process tx with Changed Threshold and Execution Success',
    async () => {
      const findings = await runTransaction('0xb58e9e81ad1dac1f33b9dfc4d19f2d909a2a3ea890c31aa27a0df10f86bd4eea')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
})
