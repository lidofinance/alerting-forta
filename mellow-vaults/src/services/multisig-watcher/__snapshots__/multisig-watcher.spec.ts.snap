// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultisigWatcher srv functional tests should process tx with Added Owner and Execution Success 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "SAFE-OWNER-ADDED",
    "description": "New owner 0x59F8D74Fe49d5ebEAc069E3baf07eB4b614BD5A7 was added to [Mellow Vaults Admin multisig](https://app.safe.global/home?safe=eth:0x9437b2a8cf3b69d782a61f9814baabc172f72003)",
    "labels": [],
    "metadata": {
      "args": "0x59F8D74Fe49d5ebEAc069E3baf07eB4b614BD5A7",
    },
    "name": "⚠️ Gnosis Safe: Owner added",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "14f4c38e-7e1e-49df-af02-6c582e79158c",
  },
  Finding {
    "addresses": [],
    "alertId": "SAFE-OWNER-ADDED",
    "description": "New owner 0xFB864C1127ff77215d0A8e652bB632F526A4A89f was added to [Mellow Vaults Admin multisig](https://app.safe.global/home?safe=eth:0x9437b2a8cf3b69d782a61f9814baabc172f72003)",
    "labels": [],
    "metadata": {
      "args": "0xFB864C1127ff77215d0A8e652bB632F526A4A89f",
    },
    "name": "⚠️ Gnosis Safe: Owner added",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "14f4c38e-7e1e-49df-af02-6c582e79158c",
  },
  Finding {
    "addresses": [],
    "alertId": "SAFE-OWNER-ADDED",
    "description": "New owner 0xa94e4ADceC38074967b17F6e8856AB12368686B5 was added to [Mellow Vaults Admin multisig](https://app.safe.global/home?safe=eth:0x9437b2a8cf3b69d782a61f9814baabc172f72003)",
    "labels": [],
    "metadata": {
      "args": "0xa94e4ADceC38074967b17F6e8856AB12368686B5",
    },
    "name": "⚠️ Gnosis Safe: Owner added",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "14f4c38e-7e1e-49df-af02-6c582e79158c",
  },
  Finding {
    "addresses": [],
    "alertId": "SAFE-OWNER-ADDED",
    "description": "New owner 0xEA685f0AF4A0BF5E7D6729b2C63eDF574D3EE27f was added to [Mellow Vaults Admin multisig](https://app.safe.global/home?safe=eth:0x9437b2a8cf3b69d782a61f9814baabc172f72003)",
    "labels": [],
    "metadata": {
      "args": "0xEA685f0AF4A0BF5E7D6729b2C63eDF574D3EE27f",
    },
    "name": "⚠️ Gnosis Safe: Owner added",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "14f4c38e-7e1e-49df-af02-6c582e79158c",
  },
  Finding {
    "addresses": [],
    "alertId": "SAFE-EXECUTION-SUCCESS",
    "description": "[TX](https://app.safe.global/transactions/tx?safe=eth:0x9437b2a8cf3b69d782a61f9814baabc172f72003&id=multisig_0x9437b2a8cf3b69d782a61f9814baabc172f72003_0x94a28231af2efbd2b1408ed0e49026adb1874fa2b91aabe5899e9123b4ce3ecb) executed by [Mellow Vaults Admin multisig](https://app.safe.global/home?safe=eth:0x9437b2a8cf3b69d782a61f9814baabc172f72003)
[blockchain explorer](https://etherscan.io/tx/0x4c3e507fd78daac24dd75e33e7c69381481db8bce62c795d2eec67126ad6d396)",
    "labels": [],
    "metadata": {
      "args": "0x94a28231af2efbd2b1408ed0e49026adb1874fa2b91aabe5899e9123b4ce3ecb,0",
    },
    "name": "ℹ️ ✅ Gnosis Safe: TX Executed",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "8dd6c439-842b-4a2a-a7e8-b404ffbac5ce",
  },
]
`;

exports[`MultisigWatcher srv functional tests should process tx with Changed Threshold and Execution Success 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "SAFE-THRESHOLD-CHANGED",
    "description": "Threshold for [Mellow Vaults Admin multisig](https://app.safe.global/home?safe=eth:0x9437b2a8cf3b69d782a61f9814baabc172f72003) was changed to 5",
    "labels": [],
    "metadata": {
      "args": "5",
    },
    "name": "⚠️ Gnosis Safe: Threshold changed",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "e70d0bb5-0033-4f34-a1cc-3b9f177f215f",
  },
  Finding {
    "addresses": [],
    "alertId": "SAFE-EXECUTION-SUCCESS",
    "description": "[TX](https://app.safe.global/transactions/tx?safe=eth:0x9437b2a8cf3b69d782a61f9814baabc172f72003&id=multisig_0x9437b2a8cf3b69d782a61f9814baabc172f72003_0xa9dc9b05a3462baba5892e962a8d91815660346c4af348e42e5d5c348a9e81a9) executed by [Mellow Vaults Admin multisig](https://app.safe.global/home?safe=eth:0x9437b2a8cf3b69d782a61f9814baabc172f72003)
[blockchain explorer](https://etherscan.io/tx/0xb58e9e81ad1dac1f33b9dfc4d19f2d909a2a3ea890c31aa27a0df10f86bd4eea)",
    "labels": [],
    "metadata": {
      "args": "0xa9dc9b05a3462baba5892e962a8d91815660346c4af348e42e5d5c348a9e81a9,0",
    },
    "name": "ℹ️ ✅ Gnosis Safe: TX Executed",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "8dd6c439-842b-4a2a-a7e8-b404ffbac5ce",
  },
]
`;
